tests/unit/mesh/test_ncmesh.cpp

```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
 
 } // test case
 
+TEST_CASE("FaceEdgeConstraint",  "[Parallel], [NCMesh]")
+{
+   constexpr int refining_rank = 0;
+   auto smesh = Mesh("../../data/ref-tetrahedron.mesh");
+
+   REQUIRE(smesh.GetNE() == 1);
+   {
+      // Start the test with two tetrahedra attached by triangle.
+      auto single_edge_refine = Array<Refinement>(1);
+      single_edge_refine[0].index = 0;
+      single_edge_refine[0].ref_type = Refinement::X;
+
+      smesh.GeneralRefinement(single_edge_refine, 0); // conformal
+   }
+
+   auto exact_soln = [](const Vector& x)
+   {
+      // sin(|| x - d ||^2) -> non polynomial but very smooth.
+      Vector d(3);
+      d[0] = -0.5; d[1] = -1; d[2] = -2; // arbitrary
+      d -= x;
+      return std::sin(d * d);
+   };
+
+   // Given a parallel and a serial mesh, perform an L2 projection and check the
+   // solutions match exactly.
+   auto check_l2_projection = [&exact_soln](ParMesh& pmesh, Mesh& smesh, int order)
+   {
+      MPI_Barrier(MPI_COMM_WORLD); // Get our ducks in a row.
```

Comment:
This test has two of these `MPI_Barrier` calls, and I confirmed they are unnecessary, by commenting them out and running the test in parallel. It is good to be cautious, but for high-level MFEM code like this test, the MPI support inside the classes should be so robust that `MPI_Barrier` is not necessary to prevent MPI disasters. Although it doesn't make any practical difference, I would prefer not to include `MPI_Barrier` calls in high-level examples like this, since it may give the impression of being necessary, which incorrectly implies that MFEM users have to be very careful about MPI in simple examples.

Reply:
Sounds good. I had originally sprinkled a few of them around during development as I was working with the `GetGlobalElementNumber` method, which if not careful does cause MPI disasters as it had a global communication. Basically had been calling it after mpi ranks had already been branched by an if. I'll get rid of them now though, as they clearly are now redundant.